src/chain.cpp

```diff
@@ -89,6 +97,9 @@ const CBlockIndex* CBlockIndex::GetAncestor(int height) const
     const CBlockIndex* pindexWalk = this;
     int heightWalk = nHeight;
     while (heightWalk > height) {
+        if (chainActive.Contains(pindexWalk->pprev))
```

Comment:
Nack, shouldn't access chainActive.

Reply:
Why not ? It is a global variable that contains the information about active chain, why we cant use it in these place. Or there are some another reasons to not use this variable?

Reply:
But this is not about the active chain, this is about the ancestor of the given block index. The goal is not to get a block at a given height in the active chain.

Reply:
Of course, but if the previous block contains in the active chain an ancestor of the current block will contains in the active chain too.  About the multiple threads, i just read the information from the variable in this case i dont think that will be some problems with it.

Reply:
It needs to lock `cs_main` which in some places we want to avoid.

Anyway, this "optimization" should be moved to the call sites, if any.

Reply:
You mean that it would be mor preferable to remove it from the GetAncestor method and replace at all places where that method calls ?

Reply:
if i'm not mistaken, if at the call site `cs_main` is locked then `GetAncestor` could be replaced by `CChain::operator[]`.

Reply:
If the previous of the current block contains in the active chain, yes we can replace.